### ISSUE TITLE
CI: Implement our own caching layer (via S3 or whatnot)

### ISSUE THREAD

**Issue by fallwith** [newrelic/newrelic-ruby-agent#986]
Recently we've been more frequently seeing GitHub Actions runs suffer cache misses that result in lengthy and even potentially error prone processes such as building Ruby from source.

For our purposes we don't particularly need caching behavior so much as a storage of prebuilt content that can be layered on top of a vanilla GitHub Action container.

Given the same base container operating system (currently we use Ubuntu 20.04) as a host, we ought to be able to prep content outside of GitHub and store it somewhere well suited for binary storage (possibly S3). From there, we can update the CI workflow to grab the prebuilt content reliably every time instead of the current dice roll based randomness we're seeing with cache.

**Comment by fallwith** [comment]
There are two key parts to this one:

- Cache the prebuilt Rubies
- Cache the needed gems

For that first part, we're already done. Since this Issue was written, we have gotten out of the business of maintaining our own Ruby build scripts and Action and have switched to using the setup-ruby Action.

For the second part, we have some complex issues standing in our way:

- setup-ruby offers `bundler-cache: true`, but for our complex CI use cases, it [doesn't work](https://github.com/newrelic/newrelic-ruby-agent/pull/1333).
- Even if we could get automatic Bundler caching to work, it would really only help out with the normal depedency-free unit tests. The multiverse suites all make use of their own separate instances of Bundler, potentially running in parallel, to process dynamically generated Gemfile files.
- Given that we have Gemfile files that won't exist until after a GitHub Actions job has kicked off a Ruby process, we'd need a way to manually access GHA cache from a running Ruby process.
- One feature-not-a-bug aspect of our current CI system is that it always brings in the latest versions of gems that we expect our users to eventually be running out on their production systems. While normal gem caching only grabs what it is available on disk, we actually want Bundler to go and see if there's a new version of something and only grab prebuilt content from cache if it's for a version we've already processed.

As per setup-ruby's [warning](https://github.com/ruby/setup-ruby#caching-bundle-install-manually), these will be "_very difficult_ to do correctly".

### Idea

- Multiverse suite processes `Envfile` to spit out multiple `Gemfile.x` files
- We perform `bundle lock` on each `Gemfile.x` file. Doing so will allow to us to stay current with the latest available gems. A `Gemfile.x.lock` file will be created, but no gems will be downloaded and no native extensions built.
- We calculate an md5 sum for the `Gemfile.x.lock` file and use it as a cache key
- If we have cache for the cache key, we download it to `$PWD/gem_cache_Gemfile.x`, as each `Gemfile.x` file will need its own cache directory.
- If we have cache, we set the `BUNDLE_CACHE_PATH` environment variable to the `Gemfile.x` specific cache dir
- If we have cache, we perform `bundle install --local`
- If we don't have cache, we run `bundle install`, and then cache the result

**Comment by fallwith** [comment]
The team has discussed this one and we've decided not to go with gem caching at this time and to consider this Issue resolved from the work done previously to address the frequent cache misses with Ruby and MySQL.

The complexity that manual caching would introduce would be too high of a cost compared to the paltry time savings we'd gain.
